[3d] Load 3D vector layer data in background + tiling

wonder-sk opened the pull request:
This adds support for background loading of data from vector layers into 3D map views. Until now, loading (and tessellation) would freeze the GUI completely - this could take many seconds depending on the complexity of input data.

The basic vector layer renderer and rule-based renderer were converted to use QgsChunkedEntity which is already used for terrain rendering. There are two more improvements in addition to unlocking of GUI:
- loading process is multi-threaded instead of using just a single core
- loading is done in tiles - so it is possible to see the tiles with 3D data appearing while other data are still being loaded

There is a new configuration option in 3D tab of vector layers - it determines how deep will be the quadtree. For example, one zoom level means there will be a single tile for the whole layer. Three zoom levels means there will be 16 tiles at the leaf level (every extra zoom level multiplies that by 4, so I have limited GUI to max. 8 levels which gives ~16K tiles which is already a lot).

How a vector layer's tiling quadtree gets populated: all internal tree nodes are empty and thus the 3D map scene tries to immediately replace them with their children - this goes until leaf nodes are reached. Only nodes at the leaf level currently hold any data. This may change in the feature when we introduce more elaborate strategies - for example, internal nodes may contain a small percentage of features of the child nodes (this would allow us to show something while zoomed out a lot, not requiring to load all data).

For debugging purposes, there is also a new configuration option "show bounding boxes". This allows you to see bounding box of each tile (especially useful if there are some issues with tiles not showing up when they should).

This work has been sponsored by a QGIS.org grant.

wonder-sk committed:
[FEATURE] Vector layers are loaded in 3D view in background

We are now using chunked entities for vector layers, so they
are split into several chunks (tiles) and each gets prepared
in background thread separately, not blocking main thread.

This is an initial implementation - work in progress.

wonder-sk committed:
Added support for background loading in 3D rule-based renderer

wonder-sk committed:
Fixed identify tool to work with chunked entities

nirvn commented:
Cool!

wonder-sk committed:
Fix crashes related to tile loading in background

- feature source must not be shared by multiple loaders (not thread safe)
- root rule must not be shared (not thread safe)
- delete map scene before map settings

wonder-sk committed:
Roll back a regression from updated poly2tri library

It looks like it was introduced when trying to make compiler happy:
https://github.com/jhasse/poly2tri/commit/e0ba327ed83f3e32933cf6cc4f61fabc50191711

Also added a test to make sure this crasher does not come back

See https://github.com/jhasse/poly2tri/issues/11 for upstream issue

wonder-sk committed:
OTF reprojection support + code style fixes

wonder-sk committed:
Introduce intermediate base class QgsAbstractVectorLayer3DRenderer

This will be useful as we start adding more properties that are common
among all vector layer 3D renderers (such as tiling configuration)

wonder-sk committed:
Add tiling configuration to vector layer 3d renderer + widget

wonder-sk committed:
sipify, doxygen, spelling

wonder-sk committed:
sip and doxygen fixes

wonder-sk committed:
Improve handling of vertical ranges

So far we used a fixed vertical range for 3D bounding boxes with range [0, 500]
This of course would not work well if Z values are outside of this range
(e.g. tiles may not show at all or get pruned prematurely)

The approach now is to start with an estimate of the vertical range based
on few features from the source vector layer and then once our tile is completely
loaded, update the 3D bounding box to better represent the actual 3D extent.

wonder-sk committed:
Optionally show bounding boxes for vector layer entities (for debugging)

wonder-sk committed:
Add information about pending jobs also from non-terrain entities

wonder-sk committed:
Use setLimit() for Z range estimation (+copy&paste error fix)

nirvn commented:
@wonder-sk , just tried this PR. **Huge* improvement to the whole 3D experience, congrats.

I've stumbled on a reproducible crash with the following gdb output:

```
Thread 1 "qgis" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe47b0e80 (LWP 30949)]
0x00007fffed33d8e4 in Qt3DCore::QNode::setEnabled(bool) () from /usr/lib/x86_64-linux-gnu/libQt53DCore.so.5
(gdb) where
#0  0x00007fffed33d8e4 in Qt3DCore::QNode::setEnabled(bool) () at /usr/lib/x86_64-linux-gnu/libQt53DCore.so.5
#1  0x00007ffff12df607 in QgsChunkedEntity::update(QgsChunkedEntity::SceneState const&) (this=0x55555bbdeb10, state=...)
    at /home/webmaster/dev/cpp/QGIS/src/3d/chunks/qgschunkedentity_p.cpp:151
#2  0x00007ffff129dec7 in Qgs3DMapScene::onFrameTriggered(float) (this=0x55555bb6c730, dt=0.0149209779) at /home/webmaster/dev/cpp/QGIS/src/3d/qgs3dmapscene.cpp:373
#3  0x00007ffff12aaa9b in QtPrivate::FunctorCall<QtPrivate::IndexesList<0>, QtPrivate::List<float>, void, void (Qgs3DMapScene::*)(float)>::call(void (Qgs3DMapScene::*)(float), Qgs3DMapScene*, void**) (f=
    (void (Qgs3DMapScene::*)(class Qgs3DMapScene * const, float)) 0x7ffff129dd2c <Qgs3DMapScene::onFrameTriggered(float)>, o=0x55555bb6c730, arg=0x7fffffffcd30)
    at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:152
#4  0x00007ffff12a9d60 in QtPrivate::FunctionPointer<void (Qgs3DMapScene::*)(float)>::call<QtPrivate::List<float>, void>(void (Qgs3DMapScene::*)(float), Qgs3DMapScene*, void**) (f=(void (Qgs3DMapScene::*)(class Qgs3DMapScene * const, float)) 0x7ffff129dd2c <Qgs3DMapScene::onFrameTriggered(float)>, o=0x55555bb6c730, arg=0x7fffffffcd30)
    at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:185
#5  0x00007ffff12a852a in QtPrivate::QSlotObject<void (Qgs3DMapScene::*)(float), QtPrivate::List<float>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x55555ba3fa60, r=0x55555bb6c730, a=0x7fffffffcd30, ret=0x0) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:414
#6  0x00007ffff33ef5c8 in QMetaObject::activate(QObject*, int, int, void**) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#7  0x00007fffe982a3e4 in Qt3DLogic::QFrameAction::triggered(float) () at /usr/lib/x86_64-linux-gnu/libQt53DLogic.so.5
#8  0x00007fffe9829fe2 in  () at /usr/lib/x86_64-linux-gnu/libQt53DLogic.so.5
#9  0x00007fffe982a07e in  () at /usr/lib/x86_64-linux-gnu/libQt53DLogic.so.5
#10 0x00007ffff3dd7a86 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#11 0x00007ffff3de0e00 in QApplication::notify(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#12 0x00007ffff4f8d6c1 in QgsApplication::notify(QObject*, QEvent*) (this=0x7fffffffd640, receiver=0x55555ba6bba0, event=0x7fff0800e2c0)
    at /home/webmaster/dev/cpp/QGIS/src/core/qgsapplication.cpp:417
#13 0x00007ffff33c3a9a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#14 0x00007ffff33c6718 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ffff341c0a7 in  () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007fffebffd84d in g_main_context_dispatch () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#17 0x00007fffebffdad0 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007fffebffdb73 in g_main_context_iteration () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#19 0x00007ffff341b6a5 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#20 0x00007ffff33c263b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#21 0x00007ffff33ca3a6 in QCoreApplication::exec() () at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#22 0x000055555556dbc3 in main(int, char**) (argc=1, argv=0x7fffffffdde8) at /home/webmaster/dev/cpp/QGIS/src/app/main.cpp:1597
```

Steps to reproduce the crash:
- Load this dataset into a new project: [buildings3d_fixed.zip](https://github.com/qgis/QGIS/files/4052618/buildings3d_fixed.zip)
- Open the style dock, go to the 3D symbology tab
- Set the zoom level to 7, activate single 3D symbol, and set the polygon's extrusion to the HEIGHT field
- Open a new 3D map canvas panel
- *boom* QGIS dies

wonder-sk commented:
@nirvn Thanks a lot for testing! Hmm that crash is caused by the fact that we currently allow up to 512 active tiles for a single entity. But here with zoom level 7 we get to a situation where the engine wants to load 4096 tiles (if zoom to the whole layer). Right now I think of just fixing it by changing the max. zoom level to 5 in order to keep max. number of tiles to 256. And later I think we will need to handle this in a more elegant way...

nirvn commented:
@wonder-sk , one more issue, UI/UX related:
![Screenshot from 2020-01-14 14-23-40](https://user-images.githubusercontent.com/1728657/72322709-9abb7100-36d9-11ea-8a1d-da475a59f3c0.png)

IMHO, we should _not_ show the zoom level counts as well as bounding box (the latter being debugging-related) by default, and all the time. It's not user friendly as it can be overwhelming for new users, and it eats vertical spacing.

My suggestion:  use a collapsible group box with a title along the lines of '3D Rendering'. That'll adopt something we also have for 2D symbology:
![Screenshot from 2020-01-14 14-23-49](https://user-images.githubusercontent.com/1728657/72322828-e40bc080-36d9-11ea-9d8d-3c039148afcd.png)

wonder-sk committed:
Fix billboard test

The test got affected by #33823 so we just side step it by moving
away from using also point at location 0,0

wonder-sk committed:
Fix 3D rule-based rendering test

wonder-sk committed:
Lower the maximum number of zoom levels to 5

With more zoom levels, we may get over the current limit of 512 active nodes
per entity and the engine would get into a vicious loop of loading tiles
and unloading other tiles (which are also needed) until it all crashes.

wonder-sk committed:
Put settings into collapsible group box as suggested by Mathieu

nirvn commented:
@wonder-sk , thanks for embracing the collapsible group box suggestion :wink: just tested the branch again, working great.

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.cpp:
```diff
@@ -0,0 +1,186 @@
+  : QgsChunkLoader( node )
+  , mFactory( factory )
+  , mContext( factory->mMap )
+  , mSource( new QgsVectorLayerFeatureSource( factory->mLayer ) )
```
is there any possibility factory == nulptr when calling this constructor?

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  QFuture<void> future = QtConcurrent::run( [req, this]
+  {
+    QgsEventTracing::ScopedEvent e( "3D", "RB chunk load" );
```
QStringLiteral

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.cpp:
```diff
@@ -0,0 +1,186 @@
+QgsRuleBasedChunkLoaderFactory::QgsRuleBasedChunkLoaderFactory( const Qgs3DMapSettings &map, QgsVectorLayer *vl, QgsRuleBased3DRenderer::Rule *rootRule, int leafLevel )
+  : mMap( map )
+  , mLayer( vl )
+  , mRootRule( rootRule->clone() )
```
can rootRule be nullptr?

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.cpp:
```diff
@@ -0,0 +1,186 @@
+{
+}
+
+QgsRuleBasedChunkLoaderFactory::~QgsRuleBasedChunkLoaderFactory()
```
=default?

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.h:
```diff
@@ -0,0 +1,119 @@
+class QgsAbstract3DSymbol;
+class QgsFeature3DHandler;
+
+#include "qgsrulebased3drenderer.h"
```
can you move to the group of other includes (few lines up)

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.h:
```diff
@@ -0,0 +1,119 @@
+    //! Creates loader for the given chunk node. Ownership of the returned is passed to the caller.
+    virtual QgsChunkLoader *createChunkLoader( QgsChunkNode *node ) const override;
+
+    const Qgs3DMapSettings &mMap;
```
should be these member variables private?

wonder-sk replied:
They are used directly from QgsRuleBasedChunkLoader so I left them public. This class is implementation detail so it's not in public api anyway...

PeterPetrik commented on src/3d/qgsrulebasedchunkloader_p.h:
```diff
@@ -0,0 +1,119 @@
+ * QgsRuleBasedChunkLoaderFactory and QgsRuleBasedChunkLoader to do the actual work
+ * of loading and creating 3D sub-entities for each tile.
+ *
+ * \since QGIS 3.12
```
not in public API,  mayber remove \ingroup and \since from these classes?

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -0,0 +1,172 @@
+  else if ( mFactory->mSymbol->type() == QLatin1String( "line" ) )
+    mHandler = Qgs3DSymbolImpl::handlerForLine3DSymbol( layer, *static_cast<QgsLine3DSymbol *>( mFactory->mSymbol.get() ) );
+  else
+    return;  // TODO: how to handle error
```
TODO

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -0,0 +1,172 @@
+
+  QSet<QString> attributeNames;
+  if ( !mHandler->prepare( mContext, attributeNames ) )
+    return;  // TODO: how to handle errors
```
TODO

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -0,0 +1,172 @@
+
+  QFuture<void> future = QtConcurrent::run( [req, this]
+  {
+    QgsEventTracing::ScopedEvent e( "3D", "VL chunk load" );
```
QStringLiterar

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.h:
```diff
@@ -0,0 +1,117 @@
+    //! Creates loader for the given chunk node. Ownership of the returned is passed to the caller.
+    virtual QgsChunkLoader *createChunkLoader( QgsChunkNode *node ) const override;
+
+    const Qgs3DMapSettings &mMap;
```
should be private:?

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.h:
```diff
@@ -0,0 +1,117 @@
+ * This loader factory is responsible for creation of loaders for individual tiles
+ * of QgsVectorLayerChunkedEntity whenever a new tile is requested by the entity.
+ *
+ * \since QGIS 3.12
```
maybe remove \ingroup and \since in private API?

wonder-sk replied:
I don't have a strong opinion on this and there is no official policy about these - I would keep it as is...

PeterPetrik commented on src/3d/qgsvectorlayerchunkloader_p.cpp:
```diff
@@ -0,0 +1,172 @@
+    mFutureWatcher->waitForFinished();
+  }
+
+  delete mHandler;
```
maybe smart pointer?

PeterPetrik commented on src/3d/symbols/qgsline3dsymbol_p.cpp:
```diff
@@ -67,10 +68,13 @@ class QgsBufferedLine3DSymbolHandler : public QgsFeature3DHandler
     {
-      QList<QgsPolygon *> polygons;
-      QList<QgsFeatureId> fids;
+      std::unique_ptr<QgsTessellator> tessellator;
```
`#include <memory>`

PeterPetrik commented on src/3d/symbols/qgsline3dsymbol_p.cpp:
```diff
@@ -131,26 +136,38 @@ void QgsBufferedLine3DSymbolHandler::processFeature( QgsFeature &f, const Qgs3DR
+  out.triangleIndexStartingIndices.append( startingTriangleIndex );
+  out.triangleIndexFids.append( fid );
+  out.tessellator->addPolygon( *polyBuffered, extrusionHeight );
+  delete polyBuffered;
```
hmm, i think the polyBuffered should be deleted in the function created it... or ideally as unique_ptr

wonder-sk replied:
eventually yes, but for the time being I would leave it not to break something...

PeterPetrik commented on src/3d/symbols/qgsline3dsymbol_p.cpp:
```diff
@@ -131,26 +136,38 @@ void QgsBufferedLine3DSymbolHandler::processFeature( QgsFeature &f, const Qgs3DR
 void QgsBufferedLine3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context, LineData &out, bool selected )
 {
-  if ( out.polygons.isEmpty() )
+  if ( !out.tessellator->dataVerticesCount() )
```
! on integer ...

PeterPetrik commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -36,6 +37,8 @@
 
 /// @cond PRIVATE
 
+class QgsTessellator;
```
both include and forward declaration, remove one

PeterPetrik commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -52,12 +55,12 @@ class QgsPolygon3DSymbolHandler : public QgsFeature3DHandler
-      QList<QgsPolygon *> polygons;
-      QList<QgsFeatureId> fids;
-      QList<float> extrusionHeightPerPolygon;  // will stay empty if not needed per polygon
+      std::unique_ptr<QgsTessellator> tessellator;
```
`#include <memory>`

PeterPetrik commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -109,10 +115,13 @@ void QgsPolygon3DSymbolHandler::processPolygon( QgsPolygon *polyClone, QgsFeatur
+  out.triangleIndexStartingIndices.append( startingTriangleIndex );
+  out.triangleIndexFids.append( fid );
+  out.tessellator->addPolygon( *polyClone, extrusionHeight );
+  delete polyClone;
```
would be better to delete in caller

PeterPetrik commented on src/3d/symbols/qgspolygon3dsymbol_p.cpp:
```diff
@@ -206,7 +218,7 @@ void QgsPolygon3DSymbolHandler::finalize( Qt3DCore::QEntity *parent, const Qgs3D
 void QgsPolygon3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const Qgs3DRenderContext &context, PolygonData &out, bool selected )
 {
-  if ( out.polygons.isEmpty() )
+  if ( !out.tessellator->dataVerticesCount() )
```
! on int

PeterPetrik commented:
1. maybe unrelated to this PR: 

causing runtime warnings `Warning: QObject::connect(Qgs3DMapScene, Qgs3DMapToolIdentify): invalid null parameter`

PeterPetrik commented:
2. unable to test since it coredumps when opening any 3d view on MacOS 10.15 somewhere in [NSOpenGLContext setView]

<img width="956" alt="Screenshot 2020-01-16 at 12 46 25" src="https://user-images.githubusercontent.com/804608/72522690-9371a000-385e-11ea-8fb4-84ed3436a9ee.png">

wonder-sk committed:
Updates based on PeterP's review

wonder-sk commented:
Something strange going on with azure pipelines - the jobs have been queued for several hours now - and when trying to look at the details of runs, I am just getting forever "Loading..." page :confused: 

I am tempted to merge without waiting for azure pipeline to do its job....
